### ISSUE TITLE
TernaryOperatorWithReturnCodeFixProvider NullReferenceException

### ISSUE THREAD

**Issue by theraot** [code-cracker/code-cracker#906]
I am getting a NullReferenceException in Visual Studio, the following is the stack pile (Spanish):

```
System.NullReferenceException : Referencia a objeto no establecida como instancia de un objeto.
   at CodeCracker.CSharp.Style.TernaryOperatorCodeFixHelper.CastToBaseType(ExpressionSyntax ifExpression,ITypeSymbol ifType,ITypeSymbol elseType,ExpressionSyntax trueExpression)
   at CodeCracker.CSharp.Style.TernaryOperatorCodeFixHelper.CreateExpressions(ExpressionSyntax ifExpression,ExpressionSyntax elseExpression,ITypeSymbol ifType,ITypeSymbol elseType,ITypeSymbol ifConvertedType,ITypeSymbol elseConvertedType,TypeSyntax typeSyntax,SemanticModel semanticModel,ExpressionSyntax& trueExpression,ExpressionSyntax& falseExpression)
   at CodeCracker.CSharp.Style.TernaryOperatorCodeFixHelper.CreateExpressions(ExpressionSyntax ifExpression,ExpressionSyntax elseExpression,SemanticModel semanticModel,ExpressionSyntax& trueExpression,ExpressionSyntax& falseExpression)
   at async CodeCracker.CSharp.Style.TernaryOperatorWithReturnCodeFixProvider.MakeTernaryAsync(<Parámetros desconocidos>)
   at async Microsoft.CodeAnalysis.CodeActions.CodeAction.GetChangedSolutionAsync(<Parámetros desconocidos>)
   at async Microsoft.CodeAnalysis.CodeActions.CodeAction.ComputeOperationsAsync(<Parámetros desconocidos>)
   at async Microsoft.CodeAnalysis.CodeActions.CodeAction.ComputePreviewOperationsAsync(<Parámetros desconocidos>)
   at async Microsoft.CodeAnalysis.CodeActions.CodeAction.GetPreviewOperationsAsync(<Parámetros desconocidos>)
   at async Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.SuggestedAction.GetPreviewResultAsync(<Parámetros desconocidos>)
   at async Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.SuggestedActionWithNestedFlavors.<>c__DisplayClass11_0.<GetPreviewAsync>b__0(<Parámetros desconocidos>)
   at async Microsoft.CodeAnalysis.Extensions.IExtensionManagerExtensions.PerformFunctionAsync[T](<Parámetros desconocidos>)
```
The following is the code to reproduce the bug:

````csharp
using System;
using System.Collections.Generic;

namespace Theraot.Core
{
    public static class ComparerExtensions
    {
        public static IComparer<T> Reverse<T>(this IComparer<T> comparer)
        {
            var originalAsReverse = comparer as ReverseComparer<T>;
            if (originalAsReverse == null)
            {
                return new ReverseComparer<T>(comparer ?? Comparer<T>.Default);
            }
            else
            {
                return originalAsReverse.Wrapped;
            }
        }

        private class ReverseComparer<T> : IComparer<T>
        {
            private readonly IComparer<T> _wrapped;

            public ReverseComparer(IComparer<T> wrapped)
            {
                _wrapped = wrapped ?? throw new ArgumentNullException(nameof(wrapped));
            }

            internal IComparer<T> Wrapped => _wrapped;

            public int Compare(T x, T y)
            {
                return _wrapped.Compare(y, x);
            }
        }
    }
}
````

The exception happens when placing the cursor on the `if` statement on line 11 and clicking the lightbulb icon. The tool adds an item without text that opens a sub menu that says `CC0013 You can use ternary operator.` Clicking on it does nothing (Although the link to https://code-cracker.github.io/diagnostics/CC0013.html works).

---

I had a look at your code to find [`CastToType`](https://github.com/code-cracker/code-cracker/blob/master/src/CSharp/CodeCracker/Style/TernaryOperatorCodeFixProvider.cs#L232). I am guessing that `ifExpression` is null.

Under that hypothesis, I see up in the stack at [`CreateExpressions`](https://github.com/code-cracker/code-cracker/blob/master/src/CSharp/CodeCracker/Style/TernaryOperatorCodeFixProvider.cs#L191), it doesn't make a null check.

I believe the culprit is somewhere in [`MakeTernaryAsync`](https://github.com/code-cracker/code-cracker/blob/master/src/CSharp/CodeCracker/Style/TernaryOperatorCodeFixProvider.cs#L32)

---

I fiddled with my code, and was able to find a situation where `TernaryOperatorWithReturnCodeFixProvider` works with my code. By changing the line:

`return new ReverseComparer<T>(comparer ?? Comparer<T>.Default);`

To

`return (IComparer <T>)new ReverseComparer<T>(comparer ?? Comparer<T>.Default);`

The code will refactor correctly. The generated code is as follows:

`return originalAsReverse == null ? (IComparer<T>)(new ReverseComparer<T>(comparer ?? Comparer<T>.Default)) : originalAsReverse.Wrapped;`

It is possible to remove the safely cast afterwards:

`return originalAsReverse == null ? new ReverseComparer<T>(comparer ?? Comparer<T>.Default) : originalAsReverse.Wrapped;`

The above makes me thing that a potential solution is to check the type of the expression that you will be placing in the return statement and add a cast to the return type of the method if it does not match. Under the assumption that code is valid, adding the cast would be redundant, but could make the refactoring work.

It is beyond me how to do that.

Edit: If it can't be done, at least don't throw an exception that causes Code-Cracker to stop working.

---

Just in case, it is relevant to reproduce the bug. I am using Microsoft Visual Studio Community 2017, .NET version is .NET 4.6.01586 and the project is a Code Library targetting .NETCoreApp 1.1, no NuGet packages or additional references added.

---

I hope to see this fixed soon.

**Comment by carloscds** [comment]
@theraot CodeCracker not support .Net Core. This will be soon!

**Comment by theraot** [comment]
@carloscds 

Repeated test on Visual Studio 2015 and Visual Studio 2017 on .NET 2.0, .NET 3.0, .NET 3.5, .NET 3.5 Client Profile,  .NET 4.0, .NET 4.0 Client Profile, .NET 4.5, .NET 4.5.1, .NET 4.5.2, .NET 4.6, .NET 4.6.1, .NET 4.6.2, Unity 3.5 .net full base Class Library, Unity 3.5 .net subset base Class Library, Unity 3.5 .net Micro Class Library, Unity 3.5 .net Web Class Library

---

Steps to reproduce:

Preconditions: Visual Studio with Code-Cracker installed

0. Open Visual Studio
1. Create a new .NET Framework Console Application
2. Copy and paste the posted code (I did it replacing the contents of Program.cs)
3. Place cursor at the start of in line 11
4. Click the lightbulb icon that appears on the left
5. Watch the notification of error that Visual Studio shows
6. [Visual Studio 2017 only] Click on the link for stack trace
7. [Visual Studio 2017 only] See that it says NullReferenceException on CastToBaseType
8. [Visual Studio 2017 only] Close Window
9. Enable again Code-Cracker
10. Go to project properties
11. If there is any .NET Framework target installet that you haven't tested, pick that (say yes to the dialog box) and go to Step 1
12. Write this comment
13. ???
14. Profit

Note: Of course, Visual Studio 2015 doesn't support C# 7. I assure that fixing the syntax errors makes no difference. You will get the same problem with the following code:

````csharp
using System;
using System.Collections.Generic;

namespace Theraot.Core
{
    public static class ComparerExtensions
    {
        public static IComparer<T> Reverse<T>(IComparer<T> comparer)
        {
            var originalAsReverse = comparer as ReverseComparer<T>;
            if (originalAsReverse == null)
            {
                return new ReverseComparer<T>(comparer ?? Comparer<T>.Default);
            }
            else
            {
                return originalAsReverse.Wrapped;
            }
        }

        private class ReverseComparer<T> : IComparer<T>
        {
            private readonly IComparer<T> _wrapped;

            public ReverseComparer(IComparer<T> wrapped)
            {
                if (wrapped == null)
                {
                    throw new ArgumentNullException(nameof(wrapped));
                }
                _wrapped = wrapped;
            }

            internal IComparer<T> Wrapped => _wrapped;

            public int Compare(T x, T y)
            {
                return _wrapped.Compare(y, x);
            }
        }
    }
}
````

I guess you don't have to take my word for it

**Comment by carloscds** [comment]
@theraot Bug confirmed!

**Comment by carloscds** [comment]
@theraot  Fixed by #911 https://github.com/code-cracker/code-cracker/pull/938
